Vault: Added custom okta provider to Vault JWT Auth (#2706)
hashicorp/web-unified-docs: content/vault/v2.x/content/docs/auth/jwt/oidc-providers/okta.mdx

You will also need to configure bound audience along the lines of
`"bound_audiences": ["api://default", "0a4........."]` if you are using the default
authorization server.

> ### Optional okta-specific configuration

Reviewer: ```suggestion
### Optional Okta-specific configuration
```
Style correction: respect 3rd party branding

If a user belongs to 100 or more groups (Okta's default authorization server group claim
limit), Okta will silently truncate the `groups` claim in the ID token. Unlike Azure AD,
Okta does not signal that truncation has occurred — there is no `_claim_names` or
`_claim_sources` pointer in the token. Vault provides Okta-specific handling to detect

> this condition and retrieve the complete group list directly from the Okta Groups API.

Reviewer: ```suggestion
Okta silently truncates the `groups` claim in the ID token for users that belong
to the default authorization server group claim limit (100 or more groups).
Unlike Azure AD, Okta does use `_claim_names` or `_claim_sources` pointers in
the token to signal when it truncates claims. Vault provides Okta-specific
handling to detect when truncation occurs and retrieve the complete group list
directly from the Okta Groups API.
```
Style correction: write in active voice, avoid possessives

#### Setup

> To set up the Okta-specific handling, you'll need:

Reviewer: ```suggestion
To set up the Okta-specific handling, you need:
```

- A dedicated Okta service account user. Create one under
  **Directory > People > Add person** in the Okta Admin Console. Using a dedicated

>   service account rather than a personal admin account is strongly recommended.

Reviewer: ```suggestion
  **Directory > People > Add person** in the Okta Admin Console. We strongly
  recommend using a dedicated service account rather than a personal admin
  account.
```
Style correction: write in active voice

- A custom admin role assigned to that service account, granting only
  **View users and their details** and **View groups and their details**. Create and
  assign custom roles under **Security > Administrators > Create new role** in the
  Okta Admin Console. The built-in **Read-Only Administrator** role also works but
  grants broader access than necessary.
- An Okta API token (SSWS) generated for that service account, under
  **Security > API > Tokens** in the Okta Admin Console. Copy the token value
  immediately after creation.

The Okta-specific handling that's used to fetch the complete group list calls the Okta
Groups API endpoint `GET /api/v1/users/{userId}/groups`, authenticated with the SSWS
API token described above. This endpoint is admin-only and cannot be reached with an
end-user OAuth access token, regardless of which `okta.*` scopes are granted to the

> OIDC application — this is a constraint of Okta's API, not a Vault limitation.

Reviewer: ```suggestion
Vault authenticates to the Okta Groups endpoint (`/api/v1/users/{userId}/groups`)
with the SSWS API token to process truncated `groups` claims. The authentication
constraint comes from Okta, not Vault. The Groups endpoint is admin-only, so
Vault cannot use an end-user OAuth access token, regardless of the `okta.*`
scopes granted. 
```
Style correction: write in active voice, avoid "this" as a pronoun, avoid possessives

  assign custom roles under **Security > Administrators > Create new role** in the
  Okta Admin Console. The built-in **Read-Only Administrator** role also works but
  grants broader access than necessary.
- An Okta API token (SSWS) generated for that service account, under
  **Security > API > Tokens** in the Okta Admin Console. Copy the token value
  immediately after creation.

Vault authenticates to the Okta Groups endpoint (`/api/v1/users/{userId}/groups`)
with the SSWS API token to process truncated `groups` claims. The authentication
constraint comes from Okta, not Vault. The Groups endpoint is admin-only, so
Vault cannot use an end-user OAuth access token, regardless of the `okta.*`
scopes granted. 

#### Configuration

- `provider` `(string: <required>)` - Name of the provider. Must be set to `"okta"`.
- `fetch_groups` `(bool: false)` - If set to `true`, Vault checks whether the `groups`
  claim in the ID token is absent or has reached `groups_cap`, and if so, fetches the
  complete group list from the Okta Groups API. When `false` (the default), Vault always
  uses the groups present in the ID token, which may be incomplete for users in 100 or

>   more groups.

Reviewer: ```suggestion
- `fetch_groups` `(bool: false)` - When `fetch_groups` is `false`, Vault always
  uses the groups present in the ID token, which may be incomplete for users in
  100 or more groups. When `fetch_groups` is `true`, Vault checks whether the
  `groups` claim in the ID token is absent or has reached `groups_cap`. If the
  claim is unusable, Vault fetches the complete group list from the Okta Groups
  API. 
```
Style correction: avoid implicit subjects/objects, write in simple sentences.

In general, we should explain the default behavior first. In this case, it also
helps explain why you might want to set the value explicitly.

- `org_url` `(string: <optional>)` - The Okta org base URL, e.g. `https://example.okta.com`.
  Must use `https`. If not provided, Vault derives it from `oidc_discovery_url` by
  using its scheme and host, stripping any custom authorization server path such as
  `/oauth2/...`. Only needs to be set explicitly if your OIDC discovery URL is on a

>   different domain than your Okta org, for example a custom vanity domain.

Reviewer: ```suggestion
- `org_url` `(string: "")` - The full Okta org base URL including `https://`. For
  example, `https://example.okta.com`. If you leave `org_url` unset, Vault
  derives the URL from `oidc_discovery_url` using the scheme and host minus any
  custom authorization server paths such as `/oauth2/...`. You only need to set
  the org URL explicitly if your OIDC discovery URL uses a vanity domain or any
  other domain that differs from your Okta org.
```
Style correction: write in active voice, use simple sentences, provide the actual default value not "option" in the parameter definition

- `api_token` `(string: <optional>)` - An Okta API token (SSWS) for the service account
  described in [Setup](#setup) above. Required when `fetch_groups` is `true`. If a
  user's group count reaches `groups_cap` and no `api_token` is configured, Vault
  returns an authentication error at login. This field is sensitive and is masked in

>   `vault read` output and audit logs.

Reviewer: ```suggestion
- `api_token` `(string: "")` - An Okta API token (SSWS) for the service account.
  Vault requires an API token when `fetch_groups` is `true`. If you leave the API
  token unset, Vault returns an authentication error at login for users that
  reach `groups_cap`. Vault treats the API token as a sensitive field and masks
  the value in audit logs and `vault read` output.
```
Style correction: avoid possessives, write in active voice, avoid implicit objects/subjects when possible

> - `user_id_claim` `(string: <optional>)` - The ID token claim whose value identifies

Reviewer: ```suggestion
- `user_id_claim` `(string: <user_claim>)` - The ID token claim whose value identifies
```

  the user in the Okta API call. Okta accepts a user ID, login (email/UPN), or unique

>   login shortname. Defaults to the role's `user_claim` when not set.

Reviewer: ```suggestion
  login shortname.
```

- `groups_cap` `(int: 100)` - The threshold at which Vault treats the `groups` claim as
  potentially truncated. If the claim is absent, or its length is greater than or equal
  to `groups_cap`, Vault fetches the complete list from the Okta API when `fetch_groups`

>   is `true`. This value should match the group claim limit configured on your Okta

Reviewer: ```suggestion
  is `true`. The cap value should match the group claim limit configured on your Okta
```

>   authorization server. Defaults to 100, which is Okta's default authorization server limit.

Reviewer: ```suggestion
  authorization server.
```
The default is already noted in the parameter definition, we don't need to repeat it here

  is `true`. The cap value should match the group claim limit configured on your Okta
  authorization server.

> - `groups_filter` `(string: <optional>)` - A Go regular expression that defines

Author: can we use `<optional>` here? Since we are not using for other fields.

> - `groups_filter` `(string: <optional>)` - A Go regular expression that defines

Reviewer: ```suggestion
- `groups_filter` `(string: "")` - A Go regular expression that defines
```
Good catch. I missed this one yesterday. It should be the empty string like the rest

  behavior is consistent regardless of which path was used. Useful when a user belongs
  to many Okta groups such as AD-sync groups, app-assignment groups, or an "Everyone"
  group but only a subset is relevant to Vault. Leave unset to pass through every group

>   Okta returns.

Reviewer: ```suggestion
- `groups_filter` `(string: <optional>)` - A Go regular expression that defines
  which group names Vault accepts for policy mapping. Vault applies the filter
  consistently to groups coming from the ID token and the Okta API. Filtering is
  useful when a user belongs to many Okta groups but only a subset is relevant
  to Vault. Leave `groups_filter` unset to pass through every group Okta returns.
```
Style correction: write in active voice, use complete sentences

Example configuration:

```shell
vault write auth/oidc/config - <<EOF
{
  "oidc_discovery_url": "https://example.okta.com/oauth2/default",
  "oidc_client_id": "your_client_id",
  "oidc_client_secret": "your_client_secret",
  "default_role": "your_default_role",
  "provider_config": {
    "provider": "okta",
    "fetch_groups": true,
    "org_url": "https://example.okta.com",
    "api_token": "00ABCxyz...",
    "groups_cap": 100,
    "user_id_claim": "sub",
    "groups_filter": "^vault-"
  }
}
EOF
```

#### Role

Vault uses the [user_claim](/vault/api-docs/auth/jwt#user_claim) value of the role as the
fallback user identifier for the Okta API lookup when you leave `user_id_claim` unset in
`provider_config`. Set it to a claim that Okta accepts as a user identifier — typically

> `sub` (Okta user ID) or `email`.

Reviewer: ```suggestion
`provider_config`. Set `user_id_claim` to a claim that Okta accepts as a user identifier, for example `sub` (Okta user ID) or `email`.
```

> fallback user identifier for the Okta API lookup whenever `user_id_claim` is not set in

Reviewer: ```suggestion
Vault uses the [user_claim](/vault/api-docs/auth/jwt#user_claim) value of the role as the
fallback user identifier for the Okta API lookup when you leave `user_id_claim` unset in
```
Style correction: write in active voice

> `provider_config`. Set it to a claim that Okta accepts as a user identifier — typically

Reviewer: What is "it" in this context? The `user_id_claim` field or the user claim?

Author: `user_id_claim`. The `user_id_claim` should be set to a value that Okta accepts as an user identifier.
